Add dataloader doc (#248)
Oneflow-Inc/libai: docs/source/tutorials/basics/Write_Dataloaders.md

# Write Dataloaders

> This tutorial explains how the dataset APIs work, and how to use them to add custom datasets.

Reviewer: ```suggestion
This tutorial explains how the dataset APIs work, and how to customize your own datasets with them.
```

# Build Common Dataloaders 

> In most cases, the default `build_nlp_train_val_test_loader`, `build_nlp_train_loader`, `build_nlp_test_loader`, `build_image_train_loader` and `build_image_test_loader` defined in [`libai/data/build.py`](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/build.py) are highly recommended to build dataloaders in LiBai.

Reviewer: ```suggestion
To build dataloaders in LiBai, we highly recommend users to use the default `build_nlp_train_val_test_loader`, `build_nlp_train_loader`, `build_nlp_test_loader`, `build_image_train_loader` and `build_image_test_loader` which are defined in [`libai/data/build.py`](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/build.py) in most cases.
```

Reviewer: 感觉这样好些

> The only thing you should do is writing `Dataset` like torch, and return `Instance` structure in `__getitem__`. The `Instance` structure stores the attributes of an instance (e.g., image, tokens) as "fields", and the `DistTensorData` structure provides a standard `to_global()`(called in `get_batch()`) function for local tensors.

Reviewer: ```suggestion
The only thing you need to do is to write pytorch style `Dataset`, and return `Instance` structure in `__getitem__`. The `Instance` structure stores the attributes of an instance (e.g., image, tokens) as "fields", and the `DistTensorData` structure provides a standard `to_global()`(called in `get_batch()`) function to convert local tensors to global tensors.
```

> The only thing you need to do is to write pytorch style `Dataset`, and return `Instance` structure in `__getitem__`. The `Instance` structure stores the attributes of an instance (e.g., image, tokens) as "fields", and the `DistTensorData` structure provides a standard `to_global()`(called in `get_batch()`) function to convert local tensors to global tensors.

Reviewer 2: 这个部分好像Instance的文档没有搞，要不要补充一下Instance的文档然后这边加个链接

Author: 这边`Instance`不用感觉单开一个doc去搞, 没什么内容,  用户也只需要用`Instance`包一下就好了

Reviewer 2: 可以的

> In `__getitem__` function, the `key` returned by the method must be consistent with the parameter name of the `forward` function in the `model`. Here is an example code: 

Reviewer: ```suggestion
The returned instance by `__getitem__` function must contains the same keys with the parameter name of the `forward` function of the `model`. The following shows an example: 
```

> The returned instance by `__getitem__` function must contain the same keys with the parameter names of the `forward` function of the `model`. The following shows an example: 

Reviewer 2: ```suggestion
The returned instance by `__getitem__` function must contain the same keys with the `args` passed in `forward` function of the `model`. The following shows an example: 
```
这样会不会好些

class MyDataset(flow.utils.data.Dataset):

    ...

    def __getitem__(self, idx):
        text = np.array(self.dataset[idx], dtype=np.long)
        # transfer to flow.tensor
        input_ids = flow.tensor(text[:-1], dtype=flow.long)
        lm_labels = flow.tensor(text[1:2], dtype=flow.long)
        # attention_mask must be a [0, 1] metric
        attention_mask = flow.tensor(text[2:3], dtype=flow.long)
        loss_mask = flow.tensor(text[3:], dtype=flow.long)
        # the keys (`input_ids` ... `labels`) should be same as the parameter name of model.forward()
        sample = Instance(
            input_ids=DistTensorData(input_ids),
            # attention_mask must be a [0, 1] metric
            attention_mask=DistTensorData(attention_mask),
            loss_mask=DistTensorData(lm_labels, placement_idx=-1),
            labels=DistTensorData(lm_labels, placement_idx=-1),
        )
        return sample

# my_model.py
import oneflow.nn as nn

class MyModel(nn.Module):
    ...
    
    # the parameters' name is the same as the returned key in __getitem__
    def forward(self, input_ids, attention_mask, loss_mask, labels):
        ...
```

> In particular, if you need to generate your own `attention_mask`, the `attention_mask` can only be `{0, 1}`. Because LiBai has already processed `attention_mask` in `libai/layers/attention.py`

Reviewer: ```suggestion
In particular, the values of `attention_mask` can only be `0` or `1` if you need to generate your own `attention_mask`. Because LiBai has already processed `attention_mask` in `libai/layers/attention.py`
```

> In particular, the values of `attention_mask` can only be `0` or `1` if you need to generate your own `attention_mask`. Because LiBai has already processed `attention_mask` in `libai/layers/attention.py`

Reviewer 2: ```suggestion
In particular, the values of `attention_mask` can only be `0` or `1` if you need to generate your own `attention_mask`. Because LiBai has already processed `attention_mask` in `libai/layers/attention.py` as follows:
```
这部分的`libai/layers/attention.py`要不要加上链接

# Write Dataloaders

> In this section, we will introduce how to implement a custom Dataloader in LiBai.

Reviewer 3: ```suggestion
This tutorial explains how the dataset APIs work, and how to use them to add custom datasets.
```

# Build Common Dataloaders 

> In most cases, we highly recommend you use default `build_nlp_train_val_test_loader`, `build_nlp_train_loader`, `build_nlp_test_loader`, `build_image_train_loader` and `build_image_test_loader` defined in [`libai/data/build.py`](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/build.py) to build dataloaders in LiBai.

Reviewer 3: ```suggestion
In most cases, the default `build_nlp_train_val_test_loader`, `build_nlp_train_loader`, `build_nlp_test_loader`, `build_image_train_loader` and `build_image_test_loader` defined in [`libai/data/build.py`](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/build.py) are highly recommended to build dataloaders in LiBai.
```

> The only thing you should do is writing `Dataset` like torch, and return `Instance` structure in `__getitem__`. In `__getitem__` function, the `key` returned by the method must be consistent with the parameter name of the `forward` function in the `model`. Here is an example code: 

Reviewer 3: 这里可以稍微用一段来解释 Instance 和 DistTensorData 是干嘛的

class MyDataset(flow.utils.data.Dataset):

    ...

    def __getitem__(self, idx):
        text = np.array(self.dataset[idx], dtype=np.long)
        # transfer to flow.tensor
        input_ids = flow.tensor(text[:-1], dtype=flow.long)
        lm_labels = flow.tensor(text[1:2], dtype=flow.long)
        # attention_mask must be a [0, 1] metric
        attention_mask = flow.tensor(text[2:3], dtype=flow.long)
        loss_mask = flow.tensor(text[3:], dtype=flow.long)
        # the keys (`input_ids` ... `labels`) should be same as the parameter name of model.forward()
        sample = Instance(
            input_ids=DistTensorData(input_ids),
            # attention_mask must be a [0, 1] metric
            attention_mask=DistTensorData(attention_mask),
            loss_mask=DistTensorData(lm_labels, placement_idx=-1),
            labels=DistTensorData(lm_labels, placement_idx=-1),
        )
        return sample

# my_model.py
import oneflow.nn as nn

class MyModel(nn.Module):
    ...
    
    # the parameters' name is the same as the returned key in __getitem__
    def forward(self, input_ids, attention_mask, loss_mask, labels):
        ...
```

> In particular, if you need to generate your own `attention_mask`, the `attention_mask` must be a [0, 1] metric. Cause LiBai has already process `attention_mask` in `libai/layers/attention.py`

Reviewer 3: [0, 1] 容易让人以为是 0-1 的闭区间
```suggestion
In particular, if you need to generate your own `attention_mask`, the `attention_mask` can only be `{0, 1}`. Because LiBai has already processed `attention_mask` in `libai/layers/attention.py`
```

```python
class MultiheadAttention(nn.Module):
    ...

    def forward(
        self,
        hidden_states: flow.Tensor,
        encoder_states: flow.Tensor = None,
        attention_mask: flow.Tensor = None,
        past_key_value: Tuple[flow.Tensor, flow.Tensor] = None,
        use_cache: bool = False,
    ):
        ...

        attention_scores = flow.matmul(query, key, transpose_b=True, alpha=self.norm_factor)

        # your passed attention_mask
        if attention_mask is not None:
            if self.scale_mask_softmax_fusion:
                attention_weights = flow._C.fused_scale_mask_softmax(
                    attention_scores, attention_mask, fill_value=-10000.0
                )
            else:
                if self.coeff is not None:
                    attention_scores *= self.coeff
                attention_scores = flow.mul(attention_scores, attention_mask)
                attention_scores = attention_scores - 10000.0 * (1 - attention_mask)

>                 attention_weights = flow.softmax(attention_scores, dim=-1)

Reviewer 3: 这里感觉只需要这3行代码就够了

        else:
            attention_weights = flow.softmax(attention_scores, dim=-1)

        attention_weights = self.dropout(attention_weights)
        context = flow.matmul(attention_weights, value)

```

> After finish your `MyDataset`, set `dataloader` in your `config.py` according to your own needs. If you only have one training dataset for nlp task and you want to split it to `train`, `valid` and `test` dataset automatically, you can choose `build_nlp_train_val_test_loader`, the evaluation will be calculated in `valid` and `test` dataset. 

Reviewer 3: ```suggestion
After finishing your `MyDataset`, set `dataloader` in your `config.py` depending on your needs. If you have only one training dataset for nlp task and want to split it into `train`, `valid` and `test` datasets automatically, you can choose `build_nlp_train_val_test_loader`, the evaluation will be calculated in `valid` and `test` dataset. 
```

> After finish your `MyDataset`, set `dataloader` in your `config.py` according to your own needs. If you only have one training dataset for nlp task and you want to split it to `train`, `valid` and `test` dataset automatically, you can choose `build_nlp_train_val_test_loader`, the evaluation will be calculated in `valid` and `test` dataset. 

Reviewer 4: After ~finish~ finishing your `MyDataset`

这里应该是finishing？

If you only have one training dataset for nlp task and ~you~ want to split it to `train`

第二个you是不是可以去掉？
